sample-programs/erc20/erc20.scm

```diff
@@ -0,0 +1,129 @@
+;;  copyright: (c) 2013-2018 by Blockstack PBC, a public benefit corporation.
+
+;;  This file is part of Blockstack.
+
+;;  Blockstack is free software. You may redistribute or modify
+;;  it under the terms of the GNU General Public License as published by
+;;  the Free Software Foundation, either version 3 of the License or
+;;  (at your option) any later version.
+
+;;  Blockstack is distributed in the hope that it will be useful,
+;;  but WITHOUT ANY WARRANTY, including without the implied warranty of
+;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+;;  GNU General Public License for more details.
+
+;;  You should have received a copy of the GNU General Public License
+;;  along with Blockstack. If not, see <http://www.gnu.org/licenses/>.
+
+;; Storage
+(define-map balances 
+  ((owner principal)) 
+  ((balance int)))
+(define-map allowances 
+  ((spender principal) (owner principal)) 
+  ((allowance int)))
+(define total-supply 0)
+
+;; Internals
+
+;; Total number of tokens in existence.
+(define (get-total-supply)
+  total-supply)
+
+;; Gets the amount of tokens owned by the specified address.
+(define (balance-of (account principal))
+  (let ((balance
+      (get balance 
+        (fetch-entry balances (tuple (owner account))))))
+    (if (eq? balance 'null) 0 balance)))
+
+;; Gets the amount of tokens that an owner allowed to a spender.
+(define (allowance-of (spender principal) (owner principal))
+  (let ((allowance (get allowance 
+      (fetch-entry allowances (tuple 
+        (owner owner) 
+        (spender spender))))))
+    (if (eq? allowance 'null) 0 allowance)))
+
+;; Credits balance of a specified principal.
+(define (credit-balance! (account principal) (amount int))
+  (if (<= amount 0)
+    'false
+    (let ((current-balance (balance-of account)))
+      (begin
+        (set-entry! balances 
+          (tuple (owner account))
+          (tuple (balance (+ amount current-balance)))) 
+        'true)))) ;; Overflow management?

```

Comment:
Numbers in the VM must be `i128`.  You should definitely be checking for overflows -- especially since this is an example code snippit that other people will learn from.

Reply:
There's a question of what's the best practice here, I think. Arithmetic errors (overflows, underflows and division by zero) all cause the transaction to abort, so if you just want the overflowed operation to fail, you don't need to do anything.

Reply:
Ok sounds good. It would be useful to have some constants - int max, int min.
The abortion behavior is great, compared to solidity (an overflow returns 0).

Reply:
However, I stand by my comment -- the sample code we show developers should nevertheless check for overflows so developers know what to expect.

Reply:
Yep -- that makes sense to me. Code that looks like this would be somewhat more pedagogical:

```
(if (eq? current-balance integer-max)
    (abort)
    (... continue)
```

 -- we'll add native constants for `integer-max` and `integer-min`.

Reply:
I might be missing something, but how can we check for overflows, if overflows are throwing?
By writing:
```
(if (eq? (+ current-balance amount) integer-max)
    (abort)
    (... continue)
```
developers might be thinking that the execution stops thanks to the `if`. 
The execution would actually stop because of the (throwing) operation (+ current-balance amount)...

Reply:
```
(<= (- integer-max current-balance) amount)
```
would do the work, but isn't it confusing?

Reply:
Yeah, _legal_ overflow checks are pretty confusing to read.